submission/persistence-filesystem/src/main/kotlin/ac/uk/ebi/biostd/persistence/filesystem/fire/FireFilesService.kt

```diff
@@ -69,31 +66,36 @@ class FireFilesService(
         )
     }
 
-    @Suppress("ReturnCount")
-    private fun reuseOrPersistFireFile(
-        file: ExtFile,
-        subRelPath: String,
-        fallbackFile: () -> File,
-    ): ExtFile {
-        val expectedPath = "/$subRelPath/${file.relPath}"
-        val files = client.findByMd5(file.md5)
-
-        val byPath = files.firstOrNull { it.filesystemEntry?.path == expectedPath }
-        if (byPath != null) return asFireFile(file, byPath.fireOid, byPath.filesystemEntry!!.path!!)
-
-        val noPath = files.firstOrNull { it.filesystemEntry?.path == null }
-        if (noPath != null) return setMetadata(noPath.fireOid, file, expectedPath)
+    private fun getOrCreate(
+        file: FireFile,
+        expectedPath: String,
+    ): FireFile {
+        return when (file.firePath) {
+            expectedPath -> file
+            null -> setMetadata(file.fireId, file, expectedPath)
+            else -> {
+                val downloaded = client.downloadByFireId(file.fireId, file.fileName)
+                val saved = client.save(downloaded, file.md5, file.size)
+                setMetadata(saved.fireOid, file, expectedPath)
+            }
+        }
+    }
 
-        val saved = client.save(fallbackFile(), file.md5, file.size)
-        return setMetadata(saved.fireOid, file, expectedPath)
+    private fun getOrCreate(file: NfsFile, expectedPath: String): FireFile {
+        val matches = client.findByMd5(file.md5)
```

Comment:
Given that we already searched for the files in the sources while verifying the request, it looks like it would be fair to assume that any NfsFile we get at this point is not on FIRE otherwise, we would have a fire file here instead. Maybe it's worth checking this out.

Reply:
it can be in FIRE, if user use NFS source , we do not check if file is fire, we may do this but it will make accepting a request super slow and application will look unresponsive for user